File: backend/app/schemas/lsc_data.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.lsc_sign import LSCSignStatus


# --- Categorías ---

class CategoryCreate(BaseModel):
    name: str = Field(min_length=2, max_length=100)
    description: str | None = Field(default=None, max_length=500)


class CategoryUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    is_active: bool | None = None


class CategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    slug: str
    description: str | None
    is_active: bool
    created_at: datetime


# --- Señas ---

class SignCreate(BaseModel):
    category_id: uuid.UUID
    word: str = Field(min_length=1, max_length=150)
    meaning: str = Field(min_length=1, max_length=500)
    description: str | None = Field(default=None, max_length=1000)
    tags: str | None = Field(default=None, max_length=300)


class SignUpdate(BaseModel):
    category_id: uuid.UUID | None = None
    word: str | None = Field(default=None, min_length=1, max_length=150)
    meaning: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = Field(default=None, max_length=1000)
    tags: str | None = Field(default=None, max_length=300)
    status: LSCSignStatus | None = None
    is_active: bool | None = None
    # Si es True, incrementa `version` (sección 12: versionar cuando sea
    # necesario). Se deja explícito para que la persona administradora
    # decida cuándo un cambio amerita una nueva versión (p. ej. corrección
    # lingüística) vs. un ajuste menor (p. ej. arreglar una errata).
    bump_version: bool = False


class SignOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    category_id: uuid.UUID
    word: str
    meaning: str
    description: str | None
    tags: str | None
    status: LSCSignStatus
    version: int
    is_active: bool
    created_at: datetime
    updated_at: datetime


class SignWithVideosOut(SignOut):
    videos: list["VideoOut"] = []
    animations: list["AnimationOut"] = []


# --- Videos ---

class VideoOut(BaseModel):
    id: uuid.UUID
    sign_id: uuid.UUID
    url: str
    original_filename: str
    content_type: str
    size_bytes: int
    is_active: bool
    created_at: datetime


# --- Animaciones del avatar 3D (Fase 6) ---

class AnimationOut(BaseModel):
    id: uuid.UUID
    sign_id: uuid.UUID
    url: str
    original_filename: str
    content_type: str
    size_bytes: int
    is_active: bool
    created_at: datetime


class TranslationItemEnriched(BaseModel):
    """
    Item de una secuencia de traducción (texto -> LSC) ya enriquecido con
    disponibilidad de animación 3D. `word`/`sign_id`/`found_in_db` vienen
    del motor NLP (app/ai/translation/nlp_engine.py, Fase 6 IA); esta clase
    vive en la capa de datos LSC para no acoplar el motor NLP a la tabla de
    animaciones — se conecta desde el servicio (ver
    LSCDataService.enrich_translation_sequence).
    """
    word: str
    sign_id: uuid.UUID | None = None
    found_in_db: bool
    # Sección 29 ("nunca inventar"): si la seña no está documentada o no
    # tiene animación cargada todavía, has_animation queda en False y el
    # frontend debe mostrar solo el subtítulo de esa palabra, nunca un
    # avatar inventado.
    has_animation: bool
    animation_url: str | None = None


class TranslationSequenceEnriched(BaseModel):
    original_text: str
    sequence: list[TranslationItemEnriched]
    # False si alguna palabra no tiene seña documentada o animación cargada
    # (sección 29): el frontend debe avisar que la traducción es incompleta.
    complete: bool


# --- Frases ---

class PhraseSignItem(BaseModel):
    sign_id: uuid.UUID
    position: int = Field(ge=0)


class PhraseCreate(BaseModel):
    text: str = Field(min_length=1, max_length=500)
    description: str | None = Field(default=None, max_length=1000)
    signs: list[PhraseSignItem] = Field(min_length=1)


class PhraseUpdate(BaseModel):
    text: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = Field(default=None, max_length=1000)
    is_active: bool | None = None
    signs: list[PhraseSignItem] | None = None


class PhraseSignOut(BaseModel):
    sign_id: uuid.UUID
    word: str
    position: int


class PhraseOut(BaseModel):
    id: uuid.UUID
    text: str
    description: str | None
    is_active: bool
    created_at: datetime
    signs: list[PhraseSignOut]


# --- Reconocimiento de LSC (Fase 5) ---

class ProcessVideoResponse(BaseModel):
    template_id: uuid.UUID
    frame_count: int
    hand_detection_frames: int


class RecognizeResponse(BaseModel):
    matched: bool
    sign_id: uuid.UUID | None = None
    word: str | None = None
    meaning: str | None = None
    confidence: float | None = None
    message: str | None = None
    closest_word: str | None = None

File: frontend/flutter_app/lib/models/lsc_data.dart
class LscCategory {
  const LscCategory({
    required this.id,
    required this.name,
    required this.slug,
    this.description,
    required this.isActive,
  });

  final String id;
  final String name;
  final String slug;
  final String? description;
  final bool isActive;

  factory LscCategory.fromJson(Map<String, dynamic> json) => LscCategory(
        id: json['id'] as String,
        name: json['name'] as String,
        slug: json['slug'] as String,
        description: json['description'] as String?,
        isActive: json['is_active'] as bool,
      );
}

enum LscSignStatus {
  draft,
  published,
  archived;

  static LscSignStatus fromJson(String value) =>
      LscSignStatus.values.firstWhere((v) => v.name == value, orElse: () => LscSignStatus.draft);

  String get label => switch (this) {
        LscSignStatus.draft => 'Borrador',
        LscSignStatus.published => 'Publicada',
        LscSignStatus.archived => 'Archivada',
      };
}

class LscSign {
  const LscSign({
    required this.id,
    required this.categoryId,
    required this.word,
    required this.meaning,
    this.description,
    this.tags,
    required this.status,
    required this.version,
    required this.isActive,
  });

  final String id;
  final String categoryId;
  final String word;
  final String meaning;
  final String? description;
  final String? tags;
  final LscSignStatus status;
  final int version;
  final bool isActive;

  factory LscSign.fromJson(Map<String, dynamic> json) => LscSign(
        id: json['id'] as String,
        categoryId: json['category_id'] as String,
        word: json['word'] as String,
        meaning: json['meaning'] as String,
        description: json['description'] as String?,
        tags: json['tags'] as String?,
        status: LscSignStatus.fromJson(json['status'] as String),
        version: json['version'] as int,
        isActive: json['is_active'] as bool,
      );
}

class LscVideo {
  const LscVideo({
    required this.id,
    required this.signId,
    required this.url,
    required this.originalFilename,
    required this.sizeBytes,
    required this.isActive,
  });

  final String id;
  final String signId;
  final String url;
  final String originalFilename;
  final int sizeBytes;
  final bool isActive;

  factory LscVideo.fromJson(Map<String, dynamic> json) => LscVideo(
        id: json['id'] as String,
        signId: json['sign_id'] as String,
        url: json['url'] as String,
        originalFilename: json['original_filename'] as String,
        sizeBytes: json['size_bytes'] as int,
        isActive: json['is_active'] as bool,
      );
}

class LscAnimation {
  const LscAnimation({
    required this.id,
    required this.signId,
    required this.url,
    required this.originalFilename,
    required this.sizeBytes,
    required this.isActive,
  });

  final String id;
  final String signId;
  final String url;
  final String originalFilename;
  final int sizeBytes;
  final bool isActive;

  factory LscAnimation.fromJson(Map<String, dynamic> json) => LscAnimation(
        id: json['id'] as String,
        signId: json['sign_id'] as String,
        url: json['url'] as String,
        originalFilename: json['original_filename'] as String,
        sizeBytes: json['size_bytes'] as int,
        isActive: json['is_active'] as bool,
      );
}

class LscSignWithVideos extends LscSign {
  const LscSignWithVideos({
    required super.id,
    required super.categoryId,
    required super.word,
    required super.meaning,
    super.description,
    super.tags,
    required super.status,
    required super.version,
    required super.isActive,
    required this.videos,
    required this.animations,
  });

  final List<LscVideo> videos;
  final List<LscAnimation> animations;

  factory LscSignWithVideos.fromJson(Map<String, dynamic> json) => LscSignWithVideos(
        id: json['id'] as String,
        categoryId: json['category_id'] as String,
        word: json['word'] as String,
        meaning: json['meaning'] as String,
        description: json['description'] as String?,
        tags: json['tags'] as String?,
        status: LscSignStatus.fromJson(json['status'] as String),
        version: json['version'] as int,
        isActive: json['is_active'] as bool,
        videos: (json['videos'] as List).map((e) => LscVideo.fromJson(e)).toList(),
        animations: (json['animations'] as List).map((e) => LscAnimation.fromJson(e)).toList(),
      );
}

class LscPhraseSignRef {
  const LscPhraseSignRef({required this.signId, required this.word, required this.position});
  final String signId;
  final String word;
  final int position;

  factory LscPhraseSignRef.fromJson(Map<String, dynamic> json) => LscPhraseSignRef(
        signId: json['sign_id'] as String,
        word: json['word'] as String,
        position: json['position'] as int,
      );
}

class LscPhrase {
  const LscPhrase({
    required this.id,
    required this.text,
    this.description,
    required this.isActive,
    required this.signs,
  });

  final String id;
  final String text;
  final String? description;
  final bool isActive;
  final List<LscPhraseSignRef> signs;

  factory LscPhrase.fromJson(Map<String, dynamic> json) => LscPhrase(
        id: json['id'] as String,
        text: json['text'] as String,
        description: json['description'] as String?,
        isActive: json['is_active'] as bool,
        signs: (json['signs'] as List).map((e) => LscPhraseSignRef.fromJson(e)).toList(),
      );
}
